source/common/ssl/context_impl.cc

```diff
@@ -12,18 +12,28 @@
 #include "common/common/hex.h"
 
 #include "fmt/format.h"
+#include "openssl/rand.h"
 #include "openssl/x509v3.h"
 
 namespace Envoy {
 namespace Ssl {
 
-const unsigned char ContextImpl::SERVER_SESSION_ID_CONTEXT = 1;
+int ContextImpl::sslContextIndex() {
```

Comment:
For my own edification, why is this needed? Was the old fixed index of 1 just broken?

Reply:
These two things are unrelated; they just happen to be nearby in the file.

The OpenSSL API let's you store arbitrary data on both an SSL_CTX and on an SSL.  You do this by registering for a global index, then you can set and get on any SSL_CTX or SSL with that global index.  This function registers for an index on first use, then returns it there-after.

We use this to store a this-pointer on the SSL_CTX so we can retrieve it in various callbacks that don't let us set an a void* context on the callback.